File: Graph_Algorithms/UnionFind/UnionFindExample.java
package Graph_Algorithms.UnionFind;

import java.util.*;

public class UnionFindExample {
    public static void main(String[] args) {
        int n = 5;
        UnionFind uf = new UnionFind(n);
        uf.union(0, 1);
        uf.union(3, 2);
        uf.union(4, 1);
        uf.union(4, 2);
        int root = uf.find(2);
        int size = uf.size(root);
        System.out.println(size);
        System.out.println(Arrays.toString(uf.id));
    }
}

class UnionFind {
    int[] id;
    int[] sz;

    // Constructor
    // id[i] -> the parent of each element (i)
    // sz[i] -> the size of a set with root (i)
    public UnionFind (int n) {
        id = new int[n];
        sz = new int[n];
        for (int i = 0; i < n; i++) {
            // the root of each element begins as itself
            id[i] = i;
            // size of each set begins at 1
            sz[i] = 1;
        }
    }

    // Find the root/ (set index) of element p
    public int find (int p) {
        int root = p;
        while (root != id[root]) {
            root = id[root];
        }
        while (p!=root) {
            int next = id[p];
            id[p] = root;
            p = next;
        }
        return root;
    }

    // Given a root, what is the size of its set?
    public int size(int p) {
        return sz[p];
    }

    // Performs a union between two elements 
    public void union(int p, int q) {
        int a = find(p),b = find(q);
        if (a==b) return;
        int s = size(a) + size(b);
        id[b] = a;
        sz[a] = s;
        sz[b] = s;
    }
}

File: Graph_Algorithms/UnionFind/UnionFindExample.py
class UnionFind:
    def __init__(self, n):
        self.parents = []
        self.sz = []
        for i in range(n):
            self.parents.append(i)
            self.sz.append(1)

    def find(self, a):
        while (a != self.parents[a]):
            a = self.parents[a]
        return a

    def union(self, a, b):
        a_root, b_root = self.find(a), self.find(b)
        if a_root == b_root:
            return
        if self.sz[a_root] < self.sz[b_root]:
            temp = a_root
            a_root = b_root
            b_root = temp
        self.parents[b_root] = a_root
        self.sz[a_root] += self.sz[b_root]
        
    def size(self, a):
        return self.sz[self.find(a)]
        
def main():
    n = 5
    uf = UnionFind(n)
    uf.union(0,1)
    uf.union(2,1)
    uf.union(3,4)
    uf.union(4,1)
    print(uf.find(0))
    print(uf.size(0))
    
if __name__ == "__main__":
    main()